src/app/linux.rs

```diff
@@ -1,20 +1,449 @@
 // vim: tw=80
-use std::error::Error;
+use std::{
+    collections::{HashMap, HashSet},
+    error::Error,
+    fmt,
+    fs,
+    io,
+    path::{Path, PathBuf},
+};
 
 use super::Snapshot;
 
-pub(super) struct SnapshotIter {}
+/// SnapshotIter allocates a vector of the paths to objsets
+/// under the desired pools on initialization.  Reading the
+/// objsets into Snapshots is lazy.
+pub(super) struct SnapshotIter {
+    idx:     usize,
+    objsets: Vec<PathBuf>,
+}
+
+/// This is the default filepath for zfs stats using ZoL.
+const DEFAULT_ZFS_STATS_PATH: &str = "/proc/spl/kstat/zfs";
 
 impl SnapshotIter {
-    pub(crate) fn new(_pool: Option<&str>) -> Result<Self, Box<dyn Error>> {
-        unimplemented!()
+    /// Simply use the default path to the stats files in procfs.  A value of `None`
+    /// for `pool` will get statistics for all pools.
+    pub(crate) fn new(pool: Option<&str>) -> Result<Self, Box<dyn Error>> {
+        Self::new_from_basepath(DEFAULT_ZFS_STATS_PATH, pool)
+    }
+
+    // Useful for testing on a mock directory.
+    fn new_from_basepath(
+        basepath: &str,
+        pool: Option<&str>,
+    ) -> Result<SnapshotIter, Box<dyn Error>> {
+        let objsets = Self::required_pools_from_basepath(basepath, pool)?
+            .into_iter()
+            .flat_map(Self::enumerate_pool)
+            .collect::<Vec<PathBuf>>();
+        Ok(SnapshotIter { idx: 0, objsets })
+    }
+
+    // Get all objset paths for a given pool.
+    // At this point all the paths should exist, because they have been
+    // checked by `required_pools_from_basepath`.  Should this somehow be
+    // expressed as an invariant?
+    fn enumerate_pool(pool: PathBuf) -> Vec<PathBuf> {
+        let is_dataset = |entry: &fs::DirEntry| {
+            let path = entry.path();
+            let name = entry.file_name();
+            let name = name.to_str().unwrap_or("");
+            path.is_file() && name.starts_with("objset")
+        };
+        fs::read_dir(pool).map_or(vec![], |dir| {
```

Comment:
Probably not a good idea to ignore errors.  I think you should return this error instead.  If your intention is to gracefully fail on systems without support, you should only do that if `Error::kind()` is `ErrorKind::NotFound`.

Reply:
Oh, the view was weird in the PR view, and I thought you were talking about the closure.  You mean we should return an error if the pool is missing.  Fair enough.